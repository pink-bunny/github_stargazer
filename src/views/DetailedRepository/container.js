import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  fetchRepositoryLang as fetchRepositoryLangActions,
} from '../../state/concepts/repository/actions';
import { repositorySelector } from '../../state/concepts/repository/selectors';
import RepositoryComponent from './component';

class Repository extends Component {
  componentDidMount() {
    const { fetchRepositoryLang, route: { params: { fullData, id } } } = this.props;

    fetchRepositoryLang(fullData, id);
  }

  render() {
    const { repository } = this.props;

    return <RepositoryComponent {...repository} />;
  }
}

Repository.defaultProps = {
  repository: null,
};

Repository.propTypes = {
  fetchRepositoryLang: PropTypes.func.isRequired,
  route: PropTypes.shape({
    params: PropTypes.shape({
      fullData: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
  repository: PropTypes.shape(),
};

const mapStateToProps = (state, { route: { params: { id } } }) => ({
  repository: repositorySelector(state, id),
});

const mapDispatchToProps = {
  fetchRepositoryLang: fetchRepositoryLangActions,
};

export default connect(mapStateToProps, mapDispatchToProps)(Repository);
