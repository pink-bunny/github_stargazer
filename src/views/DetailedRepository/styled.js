import styled from 'styled-components/native';

const ContentWrap = styled.ScrollView`
  padding: 20px 10px;
`;

const Title = styled.Text`
  font-weight: 700;
  margin-bottom: 16;
  font-size: 48;
`;

const TextWrap = styled.View`
  margin-bottom: 48;
`;

const Text = styled.Text`
  margin-bottom: 12;
  font-size: 24;
`;

export {
  Title, Text, TextWrap, ContentWrap,
};
