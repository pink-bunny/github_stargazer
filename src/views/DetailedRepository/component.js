import React from 'react';
import PropTypes from 'prop-types';
import { Icon, Button } from 'react-native-elements';
import { useNavigation } from '@react-navigation/native';

import * as S from './styled';
import toPercent from '../../utils/toPercent';
import Header from '../shared/Header';

const Repository = ({
  fullData,
  stargazersCount,
  watchersCount,
  forksCount,
  languages,
}) => {
  const navigation = useNavigation();

  return (
    <>
      <Header
        centerTitle={fullData}
        leftComponent={(
          <Button
            testID="ctaGoBack"
            onPress={() => navigation.goBack()}
            buttonStyle={{ padding: 12 }}
            icon={(
              <Icon
                name="chevron-left"
                type="octicon"
                size={18}
              />
            )}
            type="clear"
          />
        )}
      />

      <S.ContentWrap>
        <S.Title testID="detailedRepoStatsTitle">Stats</S.Title>
        <S.TextWrap testID="detailedRepoStatsList">
          <S.Text>{`Watchers: ${stargazersCount}`}</S.Text>
          <S.Text>{`Stars: ${watchersCount}`}</S.Text>
          <S.Text>{`Forks: ${forksCount}`}</S.Text>
        </S.TextWrap>

        <S.Title testID="detailedRepoLangTitle">Languages</S.Title>
        <S.TextWrap testID="detailedRepoLangList">
          {languages && toPercent(languages).map((item) => <S.Text key={item[0]}>{`${item[0]}: ${item[1]}%`}</S.Text>)}
        </S.TextWrap>
      </S.ContentWrap>
    </>
  );
};

Repository.defaultProps = {
  languages: null,
};

Repository.propTypes = {
  fullData: PropTypes.string.isRequired,
  stargazersCount: PropTypes.number.isRequired,
  watchersCount: PropTypes.number.isRequired,
  forksCount: PropTypes.number.isRequired,
  languages: PropTypes.shape(),
};

export default Repository;
