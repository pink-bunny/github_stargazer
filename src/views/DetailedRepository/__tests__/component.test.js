import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import DetailedRepository from '../component';

const mockGoBack = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    goBack: mockGoBack,
  }),
}));

describe('DetailedRepository component.', () => {
  const projectName = 'kenwheeler/slick';
  const props = {
    fullData: projectName,
    stargazersCount: 26373,
    watchersCount: 26373,
    forksCount: 5522,
    languages: {
      CSS: 17372,
      HTML: 9449,
      JavaScript: 89981,
    },
  };

  it('matches snapshot', () => {
    const { toJSON } = render(<DetailedRepository {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('Header', () => {
    const { queryByTestId } = render(<DetailedRepository {...props} />);

    fireEvent.press(queryByTestId('ctaGoBack'));
    expect(mockGoBack).toHaveBeenCalledTimes(1);

    expect(queryByTestId('headerTitle')).toHaveTextContent(projectName);
  });

  it('Content', () => {
    const { queryByTestId, getByText } = render(<DetailedRepository {...props} />);

    expect(queryByTestId('detailedRepoStatsTitle')).toHaveTextContent('Stats');
    expect(queryByTestId('detailedRepoStatsList')).toContainElement(getByText(/Watchers:\s\d+/i));
    expect(queryByTestId('detailedRepoStatsList')).toContainElement(getByText(/Stars:\s\d+/i));
    expect(queryByTestId('detailedRepoStatsList')).toContainElement(getByText(/Forks:\s\d+/i));

    expect(queryByTestId('detailedRepoLangTitle')).toHaveTextContent('Languages');
    expect(queryByTestId('detailedRepoLangList')).toContainElement(getByText(/CSS:\s\d+\.?\d?%/i));
    expect(queryByTestId('detailedRepoLangList')).toContainElement(getByText(/HTML:\s\d+\.?\d?%/i));
    expect(queryByTestId('detailedRepoLangList')).toContainElement(getByText(/JavaScript:\s\d+\.?\d?%/i));
  });
});
