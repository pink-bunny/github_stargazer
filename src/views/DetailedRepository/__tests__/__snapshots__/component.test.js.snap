// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailedRepository component. matches snapshot 1`] = `
Array [
  <View
    centerTitle="kenwheeler/slick"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "height": 64,
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingTop": 20,
        "position": "relative",
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 8,
        "zIndex": 1,
      }
    }
    testID="headerContainer"
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "borderRadius": 3,
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="ctaGoBack"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 12,
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 3,
        }
      }
    >
      <Text
        testID="headerTitle"
      >
        kenwheeler/slick
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
        }
      }
    />
  </View>,
  <View
    style={
      Array [
        Object {
          "paddingBottom": 20,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 20,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 16,
            "fontWeight": "700",
            "marginBottom": 16,
          },
        ]
      }
      testID="detailedRepoStatsTitle"
    >
      Stats
    </Text>
    <View
      style={
        Array [
          Object {
            "marginBottom": 12,
          },
        ]
      }
      testID="detailedRepoStatsList"
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 4,
            },
          ]
        }
      >
        Watchers: 26373
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 4,
            },
          ]
        }
      >
        Stars: 26373
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 4,
            },
          ]
        }
      >
        Forks: 5522
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 16,
            "fontWeight": "700",
            "marginBottom": 16,
          },
        ]
      }
      testID="detailedRepoLangTitle"
    >
      Languages
    </Text>
    <View
      style={
        Array [
          Object {
            "marginBottom": 12,
          },
        ]
      }
      testID="detailedRepoLangList"
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 4,
            },
          ]
        }
      >
        CSS: 14.9%
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 4,
            },
          ]
        }
      >
        HTML: 8.1%
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "marginBottom": 4,
            },
          ]
        }
      >
        JavaScript: 77.0%
      </Text>
    </View>
  </View>,
]
`;
